File: 01_ML/01_supervised/01_regression/01_linear_reg/python/01b_multi_var.py
"""ML: Linear Regression - Multi Variate (CLI port of ../jupyter/01b_multi_var.ipynb).

Predict house price from house area (sq.ft.) and age (years).

    y = m1*x1 + m2*x2 + b

Run:
    python 01b_multi_var.py
"""

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import linear_model

PLOT_PATH = "01b_multi_var_plot.png"


def main() -> None:
    data = {
        "area":  [2600, 3000, 3200, 3600, 4000, 4100],
        "age":   [20, 15, 18, 30, 8, 8],
        "price": [550000, 565000, 610000, 680000, 725000, 810000],
    }
    test = {
        "area": [2000, 3300, 4400],
        "age":  [10, 13, 16],
    }

    df = pd.DataFrame(data)
    print("training data:")
    print(df.to_string(index=False))

    reg = linear_model.LinearRegression()
    reg.fit(df[["area", "age"]], df["price"])

    print(f"\ncoef (area):   {reg.coef_[0]}")
    print(f"coef (age):    {reg.coef_[1]}")
    print(f"intercept:     {reg.intercept_}")

    tst = pd.DataFrame(test)
    tst["price"] = reg.predict(tst[["area", "age"]])
    print("\npredictions:")
    print(tst.to_string(index=False))

    # Regression-plane mesh over the training-data range.
    area_grid = np.linspace(df.area.min(), df.area.max(), 12)
    age_grid = np.linspace(df.age.min(), df.age.max(), 12)
    A, G = np.meshgrid(area_grid, age_grid)
    mesh = pd.DataFrame({"area": A.ravel(), "age": G.ravel()})
    Z = reg.predict(mesh).reshape(A.shape)

    fig = plt.figure(figsize=(10, 8))
    ax = fig.add_subplot(111, projection="3d")
    ax.scatter(df.area, df.age, df.price, color="magenta", marker="*", label="training")
    ax.scatter(tst.area, tst.age, tst.price, color="red", label="predicted")
    ax.plot_surface(A, G, Z, color="blue", alpha=0.3)
    ax.set_xlabel("area")
    ax.set_ylabel("age")
    ax.set_zlabel("price")
    ax.set_title("Multivariate Linear Regression in 3D")
    ax.view_init(elev=20, azim=30)
    ax.legend()
    fig.savefig(PLOT_PATH, bbox_inches="tight")
    print(f"\nWrote {PLOT_PATH}")


if __name__ == "__main__":
    main()

File: 01_ML/01_supervised/01_regression/01_linear_reg/nodejs/01b_multi_var.js
// ML: Linear Regression - Multi Variate (Node.js port of ../jupyter/01b_multi_var.ipynb)
//
// Predict house price from house area (sq.ft.) and age (years).
//
//   y = m1*x1 + m2*x2 + b
//
// Data:
//   area  | age | price
//   2600  | 20  | 550000
//   3000  | 15  | 565000
//   3200  | 18  | 610000
//   3600  | 30  | 680000
//   4000  |  8  | 725000
//   4100  |  8  | 810000

import MLR from 'ml-regression-multivariate-linear';
import { writePlotlyHtml } from './plot.js';

const data = {
  area:  [2600, 3000, 3200, 3600, 4000, 4100],
  age:   [20, 15, 18, 30, 8, 8],
  price: [550000, 565000, 610000, 680000, 725000, 810000],
};
const test = {
  area: [2000, 3300, 4400],
  age:  [10, 13, 16],
};

console.table(data.area.map((a, i) => ({ area: a, age: data.age[i], price: data.price[i] })));

const X = data.area.map((a, i) => [a, data.age[i]]);
const Y = data.price.map((p) => [p]);

const reg = new MLR(X, Y);

const m1 = reg.weights[0][0];
const m2 = reg.weights[1][0];
const b  = reg.weights[2][0];
console.log(`coef (area):  ${m1}`);
console.log(`coef (age):   ${m2}`);
console.log(`intercept:    ${b}`);

const Xtest = test.area.map((a, i) => [a, test.age[i]]);
const predictions = reg.predict(Xtest).map((row) => row[0]);

console.table(test.area.map((a, i) => ({ area: a, age: test.age[i], price: predictions[i] })));

// Build a regression-plane mesh over the training-data range to visualize the fit.
const areaGrid = linspace(Math.min(...data.area), Math.max(...data.area), 12);
const ageGrid  = linspace(Math.min(...data.age),  Math.max(...data.age),  12);
const z = ageGrid.map((g) => areaGrid.map((a) => reg.predict([[a, g]])[0][0]));

writePlotlyHtml(
  new URL('./01b_multi_var_plot.html', import.meta.url).pathname,
  'Linear Regression - Multi Variate',
  [
    {
      type: 'scatter3d',
      mode: 'markers',
      name: 'training',
      x: data.area,
      y: data.age,
      z: data.price,
      marker: { color: 'magenta', symbol: 'diamond', size: 6 },
    },
    {
      type: 'scatter3d',
      mode: 'markers',
      name: 'predicted',
      x: test.area,
      y: test.age,
      z: predictions,
      marker: { color: 'red', size: 6 },
    },
    {
      type: 'surface',
      name: 'fit plane',
      x: areaGrid,
      y: ageGrid,
      z,
      opacity: 0.5,
      showscale: false,
      colorscale: 'Blues',
    },
  ],
  {
    scene: {
      xaxis: { title: 'area' },
      yaxis: { title: 'age' },
      zaxis: { title: 'price' },
      camera: { eye: { x: 1.6, y: 1.6, z: 0.6 } },
    },
  },
);

console.log('\nWrote 01b_multi_var_plot.html — open it in a browser.');

function linspace(start, stop, n) {
  if (n < 2) return [start];
  const step = (stop - start) / (n - 1);
  return Array.from({ length: n }, (_, i) => start + step * i);
}
